src/Features/Core/Portable/Shared/Naming/IdentifierNameParts.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles;
+using Microsoft.CodeAnalysis.NamingStyles;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Shared.Naming
+{
+    internal readonly struct IdentifierNameParts
+    {
+        public readonly string BaseName;
+        public readonly ImmutableArray<string> BaseNameParts;
+
+        public IdentifierNameParts(string baseName, ImmutableArray<string> baseNameParts)
+        {
+            BaseName = baseName;
+            BaseNameParts = baseNameParts;
+        }
+
+        public static IdentifierNameParts GetIdentifierBaseName(ISymbol symbol, ImmutableArray<NamingRule> rules)
+        {
+            var baseName = RemovePrefixesAndSuffixes(symbol, rules, symbol.Name);
+
+            var parts = StringBreaker.GetWordParts(baseName);
```

Comment:
There's also technically the concept of a rule-specific word-separator character. But I think `StringBreaker` handles `_` correctly which is pretty much the only reasonable such character, so this is probably fine.

Reply:
I also considered that and came to the same conclusion.  If we get requests to handle other characters for string breakers we can do so, but it seems unlikely.